### PR TITLE
Running window optimization for `LAST()`

### LINKED ISSUE #9299: [FEA] Add running window fixer for first and last
**Is your feature request related to a problem? Please describe.**
Running window, unbounded preceding to current row, is a very common operation. First and last do show up here, but not that often. However, they should be very trivial to support so we should do it. Just make sure to take into account the null-ability of first and last.

### LINKED ISSUE #9520: [FEA] Add support for `LAST()` as running window function
Follow-up to #9299, and #9489.

The `LAST()` window function currently requires a single batch. This could lead to OOMs. It should be possible to implement `LAST()` using a running-window fixer:
1. When `ignoreNulls == true`, simply return the current row.
2. When `!ignoreNulls`, fix up with the last non-null result for that partition key combination.

This can follow the same trajectory as did `FIRST()`. Note that a double-pass solution is not required for `LAST()`, because:
1. The scope is only for `[UNBOUNDED_PRECEDING, CURRENT ROW]`.
2. The range-query case will be addressed by moving the batch boundaries to consider the order-by columns. In doing so, batches will include all similar rows of the order-by column, for any given partition-key combination.

### PR DESCRIPTION
**Pull request by mythrocks**
Fixes #9520.
Fixes #9299.

This is a followup to the changes in #9489, which adds the running window optimization to the `FIRST()` window fuction.

This commit adds running window support for the `LAST()` window function, with support for both `ROWS` and `RANGE` based window specifications.
This change should allow the `LAST()` aggregation to run in multiple batches if the window is `[UNBOUNDED PRECEDING, CURRENT ROW`]. This will allow for much larger window and group sizes, because the group no longer needs to fit in GPU memory. This should help mitigate out-of-memory errors with `LAST()` window aggregations.

### CODE REVIEW

**Comment by mythrocks** [comment]
What's up so far is an initial stab. It should be possible to move the commonality in `FirstRunningWindowFixer` and `LastRunningWindowFixer` into a common place, and reduce repetition.

I'm working on polishing this now.

**Comment by mythrocks** [comment]
Build

**Comment by mythrocks** [comment]
Build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala [hunk 1]
```diff
@@ -1245,6 +1245,127 @@ class FirstRunningWindowFixer(ignoreNulls: Boolean = false)
...
+          // No rows in the same partition. Current result needs no fixing.
+          incRef(unfixedWindowResults)
+        }
+      case (Some(prev), Left(someRowsInSamePartition)) => // Boolean vector.
+        if (!ignoreNulls || !prev.isValid) {
+          // If !ignoreNulls, current result needs no fixing. The latest answer is the right one.
+          // If ignoreNulls, but prev is NULL, current result is again the right answer.
+          incRef(unfixedWindowResults)
+        }
+        else {
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala:1321]: nit:

```
} else {
```

**Review by revans2 — APPROVED** [review]

**Review by revans2 — APPROVED** [review]

**Comment by mythrocks** [comment]
Added some minor refactor, for readability in the `match{}` clauses. This caused @revans2's review to be dismissed, again.  Apologies.

This might be for the best.  We can leave this where it stands to avoid build breakage over the weekend.

**Comment by mythrocks** [comment]
bUiLd

**Comment by mythrocks** [comment]
Build

**Review by revans2 — APPROVED** [review]

**Pull request merged**